runtime/interpreter/interpreter_transaction.go

```diff
@@ -148,6 +173,74 @@ func (interpreter *Interpreter) declareTransactionEntryPoint(declaration *ast.Tr
 }
 
 func (interpreter *Interpreter) VisitTransactionRoleDeclaration(_ *ast.TransactionRoleDeclaration) StatementResult {
-	// TODO:
-	panic("TODO")
+	panic(errors.NewUnreachableError())
+}
+
+func (interpreter *Interpreter) declareTransactionRole(
+	declaration *ast.TransactionRoleDeclaration,
+) (
+	roleValue *SimpleCompositeValue,
+	prepareFunctionValue *HostFunctionValue,
+) {
+	transactionRoleType := interpreter.Program.Elaboration.TransactionRoleDeclarationType(declaration)
+
+	var prepareFunction *ast.FunctionDeclaration
+	var prepareFunctionType *sema.FunctionType
+	if declaration.Prepare != nil {
+		prepareFunction = declaration.Prepare.FunctionDeclaration
+		prepareFunctionType = transactionRoleType.PrepareFunctionType()
+	}
+
+	staticType := NewCompositeStaticTypeComputeTypeID(
+		interpreter,
+		interpreter.Location,
+		fmt.Sprintf(".%s", declaration.Identifier.Identifier),
+	)
+
+	roleValue = NewSimpleCompositeValue(
+		interpreter,
+		staticType.TypeID,
+		staticType,
+		nil,
+		map[string]Value{},
+		nil,
+		nil,
+		nil,
+	)
+
+	// Construct a raw HostFunctionValue without a type,
+	// instead of using NewHostFunctionValue, which requires a type.
+	//
+	// This host function value is an internally created and used function,
+	// and can never be passed around as a value.
+	// Hence, the type is not required.
+
+	common.UseMemory(interpreter, common.HostFunctionValueMemoryUsage)
+
+	prepareFunctionValue = &HostFunctionValue{
+		Function: func(invocation Invocation) Value {
+			interpreter.activations.PushNewWithCurrent()
```

Comment:
I guess the activation retrieved below (i.e: `interpreter.activations.CurrentOrNew()`) is always the one that is pushed here?
So maybe directly use the returned:
```go
transactionRoleScope := interpreter.activations.PushNewWithCurrent()
```

Reply:
Good point, improved in 55a572f54713dc2abc638ce34b2966ce497f43db